boards: stm: arduino_portenta_h7: support onboard ethernet + pf1550 driver

facchinm opened the pull request:
This PR adds support for the onboard ethernet phy on Arduino Portenta H7.
To properly provide the 1V2 rail for the PHY, I also implemented a driver for the PF1550 PMIC (nice side effect :slightly_smiling_face: ).

The driver is feature complete for our factory OTP programming on the A0 variant. Other variants are untested (the only difference would be BUCK1/2 voltage ranges)

~On the charger side, I'd need some guidance on how to expose the led configuration via device tree (at the moment it's hardcoded in the init function~

facchinm committed:
portenta_h7: fix dangerous misconfiguration

Selecting the wrong power scheme could potentially destroy the board.
Luckily, the bit can only be set once and the default build still
uses the Arduino bootloader (which has the correct setting).

Signed-off-by: Martino Facchin <m.facchin@arduino.cc>

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+#include <zephyr/sys/linear_range.h>
+
+#include "zephyr/logging/log.h"
+LOG_MODULE_REGISTER(pf1550_charger);
```
use `LOG_MODULE_REGISTER(pf1550_charger, CONFIG_CHARGER_LOG_LEVEL);`, this way you can control the log level with the subsystem option

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+
+#define INT_ENABLE_DELAY K_MSEC(500)
+
+enum {
+  /* Charger Register Addresses */
+  CHARGER_CHG_INT           = 0x80 + 0x00,
+  CHARGER_CHG_INT_MASK      = 0x80 + 0x02,
+  CHARGER_CHG_INT_OK        = 0x80 + 0x04,
+  CHARGER_VBUS_SNS          = 0x80 + 0x06,
+  CHARGER_CHG_SNS           = 0x80 + 0x07,
+  CHARGER_BATT_SNS          = 0x80 + 0x08,
+  CHARGER_CHG_OPER          = 0x80 + 0x09,
+  CHARGER_CHG_TMR           = 0x80 + 0x0A,
+  CHARGER_CHG_EOC_CNFG      = 0x80 + 0x0D,
+  CHARGER_CHG_CURR_CFG      = 0x80 + 0x0E,
+  CHARGER_BATT_REG          = 0x80 + 0x0F,
+  CHARGER_BATFET_CNFG       = 0x80 + 0x11,
+  CHARGER_THM_REG_CNFG      = 0x80 + 0x12,
+  CHARGER_VBUS_INLIM_CNFG   = 0x80 + 0x14,
+  CHARGER_VBUS_LIN_DPM      = 0x80 + 0x15,
+  CHARGER_USB_PHY_LDO_CNFG  = 0x80 + 0x16,
+  CHARGER_DBNC_DELAY_TIME   = 0x80 + 0x18,
+  CHARGER_CHG_INT_CNFG      = 0x80 + 0x19,
+  CHARGER_THM_ADJ_SETTING   = 0x80 + 0x1A,
+  CHARGER_VBUS2SYS_CNFG     = 0x80 + 0x1B,
+  CHARGER_LED_PWM           = 0x80 + 0x1C,
+  CHARGER_FAULT_BATFET_CNFG = 0x80 + 0x1D,
+  CHARGER_LED_CNFG          = 0x80 + 0x1E,
+  CHARGER_CHGR_KEY2         = 0x80 + 0x1F,
+};
```
usually in zephyr code register offets are in `#defines` rather than an enum, it's not like you are using the enum as an argument or in a switch anyway

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+
+static int pf1550_get_charger_online(const struct device *dev, enum charger_online *online)
+{
+	enum {
+		PF1550_CHARGER_OFF_LINEAR_OFF,
+		PF1550_CHARGER_OFF_LINEAR_ON,
+		PF1550_CHARGER_ON_LINEAR_ON,
+	};
```
just use a define for `PF1550_CHARGER_ON_LINEAR_ON` outside of here?

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+}
+
+static int pf1550_set_constant_charge_current(const struct device *dev,
+						uint32_t current_ua)
```
indentation looks off

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+
+	return i2c_reg_update_byte_dt(&config->bus,
+				      CHARGER_CHG_CURR_CFG,
+				      GENMASK(4,0),
```
use a define for the mask here, same for the others

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+	uint8_t val;
+	int ret;
+
+	ret = linear_range_group_get_index(charger_battery_termination_uv_range, ARRAY_SIZE(charger_battery_termination_uv_range), voltage_uv, &idx);
```
this could probably be wrapped, compliance should complain after 100 columns

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+{
+	const struct charger_pf1550_config *config = dev->config;
+	uint8_t dummy;
+	uint8_t *int_src;
+
+	int_src = (int_a != NULL) ? int_a : &dummy;
```
I would flip this around, always read into dummy and then `if (int_a = NULL) {*int_a = dummy;}`

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+
+static int pf1550_enable_interrupts(const struct device *dev)
+{
+	enum {MASKA_VAL_ENABLE = 0xFF};
```
add a define outside of the function for these ones

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+
+	struct charger_pf1550_led_config* cfg = data->led_config;
+
+	val = (cfg->enabled ? BIT(7) : 0) | (cfg->duty / 3);
```
add a define for `BIT(7)` and the others below

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+	return i2c_reg_write_byte_dt(&config->bus, CHARGER_LED_CNFG, val);
+}
+
+#define LED_FREQUENCY_1_HZ			0
+#define LED_FREQUENCY_0_5_HZ		1
+#define LED_FREQUENCY_256_HZ		2
+#define LED_FREQUENCY_8_HZ			3
```
indentation is off?

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+							  gpio_cb);
+	int ret;
+
+	(void) pf1550_enable_interrupt_pin(data->dev, false);
```
nit: `(void)pf1550...` (drop the space)

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+
+	ret = pf1550_get_charger_status(data->dev, &data->charger_status);
+	if (ret < 0) {
+		LOG_WRN("Failed to read charger status: %d", ret);
```
do you really want to try and go ahead with the other calls here? if `pf1550_get_charger_status` likely the whole i2c bus is hosed and there's no hope for any of the other calls to succseed ever, may be better off just `return ret` and keep the rest of the code inline

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct charger_pf1550_data *data = CONTAINER_OF(dwork, struct charger_pf1550_data,
+							  int_enable_work);
```
indentation is off

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+
+	ret = gpio_pin_configure_dt(&config->int_gpio, GPIO_INPUT);
+	if (ret < 0) {
+		LOG_ERR("Could not configure interrupt GPIO");
```
log the error code

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+	gpio_init_callback(&data->gpio_cb, pf1550_gpio_callback, BIT(config->int_gpio.pin));
+	ret = gpio_add_callback_dt(&config->int_gpio, &data->gpio_cb);
+	if (ret < 0) {
+		LOG_ERR("Could not add interrupt GPIO callback");
```
same

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+};
+
+/*
+#define PF1550_LED_DEFINE(inst)								\
+	static struct charger_pf1550_led_config_##inst = {				\
+		.enabled = DT_INST_PROP(inst, constant_charge_current_max_microamp),	\
+	}
+
+	TODO: add led_config field to dt
```
yeah I think getting the LED configuration from the devicetree is fine, just put some enums with the possible options

fabiobaltieri commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+*/
+
+#define PF1550_DEFINE(inst)									\
+	static struct charger_pf1550_data charger_pf1550_data_##inst;			\
```
the `\` are a bit all over the place, pick a style and stick to it (spaces, tabs, whatever)

fabiobaltieri commented on drivers/mfd/mfd_pf1550.c:
```diff
@@ -0,0 +1,51 @@
+	return 0;
+}
+
+#define MFD_PF1550_DEFINE(inst)					\
```
`\` are off

fabiobaltieri commented on drivers/regulator/Kconfig.pf1550:
```diff
@@ -0,0 +1,29 @@
+	  Init priority for the NXP PF1550 regulator driver
+	  (common part). It must be greater than I2C and MFD init priority.
+
+config REGULATOR_NXP_PF1550_INIT_PRIORITY
+	int "PF1550 regulator driver init priority"
+	default 87
+	help
+	  Init priority for the NXP PF1550 regulator driver. It must be
+	  greater than REGULATOR_NXP_PF1550_COMMON_INIT_PRIORITY
```
if you use the same priority value on all devices they should all get sequence correctly thanks to the subpriority value, want to try that? may get rid of few symbols, I would keep the MFD one for all of them

fabiobaltieri reviewed (commented):
Good stuff, check your compliance error as well (you can also run them locally `./scripts/ci/check_compliance.py -c ...`, also the commits need al la description and signed-off-by

rriveramcrus commented:
@facchinm, just a helpful tip, quite a few of the indentation and line wrapping issues can be fixed with a `git clang-format` :)

rriveramcrus commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,48 @@
+    type: int
+    required: true
+
+  battery-termination-microvolt:
```
Looking at the datasheet this is the charge regulation voltage. The `const-charge-voltage-max-microvolt` property should be used.

rriveramcrus commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,695 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/linear_range.h>
+
+#include "zephyr/logging/log.h"
```
I think this should be `#include <zephyr/logging/log.h>`

facchinm committed:
drivers: add nxp pf1550 mfd (charger+regulator)

Add driver for nxp pf1550 PMIC

Signed-off-by: Martino Facchin <m.facchin@arduino.cc>

facchinm committed:
portenta_h7: enable onboard regulator

Enable PF1550 PMIC for Arduino Portenta H7

Signed-off-by: Martino Facchin <m.facchin@arduino.cc>

facchinm committed:
portenta_h7: enable ethernet driver

On older Portenta H7, 1V2 power rail must be enabled
to get a functional ethernet phy.

Signed-off-by: Martino Facchin <m.facchin@arduino.cc>

rriveramcrus commented on drivers/charger/charger_pf1550.c:
```diff
@@ -0,0 +1,692 @@
+	const struct charger_pf1550_config *const config = dev->config;
+
+	data->charger_enabled = enable;
+
```
Your cached state may be incorrect if the subsequent `i2c_reg_update_byte_dt` transaction fails.

facchinm replied:
Fixed, thx

gmarull commented on drivers/mfd/mfd_pf1550.c:
```diff
@@ -0,0 +1,50 @@
+	struct i2c_dt_spec bus;
+};
+
+static int mfd_pf1550_init(const struct device *dev)
+{
+	const struct mfd_pf1550_config *config = dev->config;
+	uint8_t val;
+	int ret;
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		return -ENODEV;
+	}
+
+	ret = i2c_reg_read_byte_dt(&config->bus, PF1550_REG_CHIP_ID, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (val != PF1550_CHIP_ID_VAL) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
```
this looks pretty much redundant, is an MFD driver instance really needed to just check chip ID?

facchinm replied:
Other "trivial" MFD drivers are doing the same, I'd keep it for consistency

gmarull commented on drivers/regulator/regulator_pf1550.c:
```diff
@@ -0,0 +1,442 @@
+#include <zephyr/kernel.h>
+#include <zephyr/sys/linear_range.h>
+
+#define PMIC_DEVICE_ID         (0x00)
+#define PMIC_OTP_FLAVOR        (0x01)
+#define PMIC_SILICON_REV       (0x02)
+#define PMIC_INT_CATEGORY      (0x06)
```
Parenthesis redundant

gmarull commented on drivers/regulator/regulator_pf1550.c:
```diff
@@ -0,0 +1,442 @@
+	return regulator_pf1550_set_enable(dev, false);
+}
+
+static int regulator_pf1550_set_mode(const struct device *dev, regulator_mode_t mode)
+{
+	/*
+	 * TODO: support LOAD SWITCH mode
+	 * BIT 4 of LDO1/3_CTRL
+	 */
+	return -ENOTSUP;
+}
```
remove, impl is not mandatory (API returns -ENOSYS)

facchinm replied:
I would leave it there so a future contributor can complete the implementation :slightly_smiling_face:

gmarull commented on drivers/regulator/regulator_pf1550.c:
```diff
@@ -0,0 +1,442 @@
+	uint16_t idx;
+	int ret;
+
+	ret = linear_range_group_get_win_index(range, nranges, min_uv, max_uv, &idx);
+	if (ret == -EINVAL) {
```
only EINVAL handled here?

facchinm replied:
Fixed, thx :slightly_smiling_face:

gmarull commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,48 @@
+
+properties:
+  constant-charge-current-max-microamp:
+    required: true
+    type: int
+
+  vbus-current-limit-microamp:
+    type: int
+    required: true
+
+  constant-charge-voltage-max-microvolt:
+    type: int
+    required: true
+
+  system-voltage-min-threshold-microvolt:
+    type: int
+    required: true
+    enum:
+      - 3500000
+      - 3700000
+      - 4300000
+    description: |
+      System voltage minimum threshold.
+
+  thermistor-monitoring-mode:
+    type: string
+    required: true
+    enum:
+      - "disabled"
+      - "thermistor"
+      - "JEITA-1"
+      - "JEITA-2"
+    description: |
+      Thermistor monitoring mode.
+      Refer to ThrmCfg register description and Table 2 for details.
+
+  int-gpios:
+    type: phandle-array
+    required: true
+    description: Interrupt pin
```
please, prefix vendor specific properties, eg `nxp,prop-name`.

gmarull commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7_defconfig:
```diff
@@ -35,6 +35,14 @@ CONFIG_UART_LINE_CTRL=y
+
+# Enable charger
+CONFIG_CHARGER=y
+CONFIG_CHARGER_PF1550=y
```
this should not be needed, you just enable class-level Kconfig options

gmarull commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7_defconfig:
```diff
@@ -35,6 +35,14 @@ CONFIG_UART_LINE_CTRL=y
 # Enable regulator
 CONFIG_REGULATOR=y
 CONFIG_REGULATOR_FIXED=y
+CONFIG_REGULATOR_PF1550=y
```
same

facchinm replied:
Fixed

gmarull commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7_defconfig:
```diff
@@ -35,6 +35,14 @@ CONFIG_UART_LINE_CTRL=y
 # Enable USB Stack
 CONFIG_USB_DEVICE_STACK=y
+
+# Enable i2c driver
+CONFIG_I2C=y
```
not needed

gmarull commented on boards/arduino/portenta_h7/Kconfig.defconfig:
```diff
@@ -3,6 +3,9 @@
 
 if BOARD_ARDUINO_PORTENTA_H7
 
+config REGULATOR
+	default y if NETWORKING
```
this looks wrong: device driver relying on regulator to be used should select this option

gmarull replied:
Networking does not directly depend on regulators, but some device driver. This driver should be the one selecting regulator enablement. The reality though is that regulator usage is nowadays not great, we have no infrastructure for automatically turning them on/off on a device level. So maybe the defconfig should be linked to the device itself, at least future refactors will be easier to spot what's going on.

facchinm commented:
Sorry @gmarull and @rriveramcrus for the late response, I fixed most of the issues pointed out.
I commented inline on a couple of questions I still have; hope they are not blocking for merge.

fabiobaltieri commented:
@facchinm few pending comments still, check the commits tab

gmarull commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,69 @@
+    description: |
+      Behaviour for charger LED.
+
+  #  The following macros are available:
+  #  - LED_FREQUENCY_1_HZ
+  #  - LED_FREQUENCY_0_5_HZ
+  #  - LED_FREQUENCY_256_HZ
+  #  - LED_FREQUENCY_8_HZ
+  pf1550,led-frequency-duty:
+    type: array
+    default: [0, 0]
+    description: |
+      Frequency and duty cycle for charger LED.
```
please, allow users to describe properties using SI units in DT, also, use vendor prefix not part. Example: `nxp,led-frequency-mhz = <500>`.

facchinm replied:
Only that 4 frequencies are available, adding a rounding logic would make everything extremely confusing.

gmarull replied:
Enum can be used to limit choices

facchinm replied:
@gmarull (newbie question) how do you achieve it?

gmarull replied:
actually, why is property an array? if it is like this to specify 2 things (freq + duty), this should be split into 2 integer properties, then, use `enum` to limit frequency choices (must be in `millihz` as it seems to support fractional values, DT does not support floats)

gmarull commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,69 @@
+    description: |
+      System voltage minimum threshold.
+
+  thermistor-monitoring-mode:
```
vendor prefix missing

facchinm replied:
About the prefix, all the other drivers (like https://github.com/zephyrproject-rtos/zephyr/blob/ae20a6aa9148ad16f8f100bb9ed57b4ec7a72ff2/dts/bindings/charger/ti%2Cbq25180.yaml#L22-L44 ) are not using them. Is it something that got enforced recently?

rriveramcrus replied:
I'm not sure the vendor prefix is necessary on the properties inherited from `battery.yaml` as they are generic and vendor agnostic.

gmarull replied:
if inherited, type should not be here then...

gmarull commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,69 @@
+    type: int
+    required: true
+
+  system-voltage-min-threshold-microvolt:
```
prefix

gmarull commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,69 @@
+compatible: "nxp,pf1550-charger"
+
+properties:
+  constant-charge-current-max-microamp:
+    required: true
+    type: int
+
+  vbus-current-limit-microamp:
+    type: int
+    required: true
+
+  constant-charge-voltage-max-microvolt:
```
same

rriveramcrus commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7.dts:
```diff
@@ -85,6 +86,55 @@
+			thermistor-monitoring-mode = "thermistor";
+			int-gpios = <&gpiok 0 0>;
+			pf1550,led-behaviour = <ON_DURING_CHARGING_FLASH_DURING_FAULT>;
+			/* pf1550,led-frequency-duty = <LED_FREQUENCY_8_HZ 50>; */
```
Is this commenting out of `pf1550,led-frequency-duty` intentional? 
If so, could you add a note explaining the intention?

rriveramcrus reviewed (changes requested):
Still some unresolved comments on the charger driver itself as @fabiobaltieri mentioned.

facchinm commented:
All the change requests should be addressed now thanks to @pillo79 :slightly_smiling_face:

pillo79 commented:
@gmarull concerns have been addressed, please revisit your -1. Thanks!

gmarull commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7_defconfig:
```diff
@@ -25,3 +25,9 @@ CONFIG_UART_LINE_CTRL=y
 CONFIG_REGULATOR=y
 CONFIG_REGULATOR_FIXED=y
+
+# Enable charger
+CONFIG_CHARGER=y
+
+# Enable USB Stack
+CONFIG_USB_DEVICE_STACK=y
```
boards need to enable the minimum, e.g. to run hello world, is charger/usb stack needed there? If not, drop

gmarull commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7.dts:
```diff
@@ -9,6 +9,7 @@
 #include <st/h7/stm32h747xihx-pinctrl.dtsi>
 #include "arduino_portenta_h7-common.dtsi"
 #include <../boards/common/usb/cdc_acm_serial.dtsi>
+#include <dt-bindings/charger/pf1550.h>
```
```suggestion
#include <zephyr/dt-bindings/charger/pf1550.h>
```

gmarull commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,58 @@
+    required: true
+    description: Interrupt pin
+
+  #  The following macros are available:
+  #  - LED_ON_IN_CHARGING_FLASH_IN_FAULT
+  #  - LED_FLASH_IN_CHARGING_ON_IN_FAULT
+  #  - LED_MANUAL_OFF
+  pf1550,led-behaviour:
```
should likely be a string property, see e.g. how nPM1300 LED behavior is specified. You can later use DT_ENUM_IDX helpers to get a number.

gmarull reviewed (commented):
a few minor nits, otherwise lgtm

rriveramcrus commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,57 @@
+  pf1550,led-behaviour:
+    type: string
+    enum:
+    - on_in_charging_flash_in_fault
+    - flash_in_charging_on_in_fault
+    - manual_off
```
These need to be indented another level to satisfy the compliance checks and have a valid dt-binding.

Additionally these strings should be in quotes as is done for the `pf1550,thermistor-monitoring-mode` string enum.

pillo79 replied:
Thanks for the heads-up, updated (with quotes for good measure :slightly_smiling_face:)

rriveramcrus reviewed (changes requested):
Just needs the compliance fix to get my +1 again :)

rriveramcrus reviewed (dismissed):
Thank you for your contribution!

pillo79 commented:
@gmarull please revisit your -1, thanks!

gmarull commented on dts/bindings/charger/nxp,pf1550-charger.yaml:
```diff
@@ -0,0 +1,58 @@
+    type: string
+    required: true
+    enum:
+      - "on_in_charging_flash_in_fault"
+      - "flash_in_charging_on_in_fault"
+      - "manual_off"
```
```suggestion
      - "on-in-charging-flash-in-fault"
      - "flash-in-charging-on-in-fault"
      - "manual-off"
```
